File: python/VMPrototype1/frame.py
from instructions import Instructions
from objects import AdClosureObject, AdCompiledInstance


class Frame:
    def __init__(self, cl: AdClosureObject, ip: int, base_pointer: int = 0, bound_instance: AdCompiledInstance = None):
        self.cl = cl
        self.ip = ip
        self.base_pointer = base_pointer
        self.bound_instance = bound_instance

    def instructions(self) -> Instructions:
        return self.cl.fn.instructions


def new_frame(cl: AdClosureObject, base_pointer: int) -> Frame:
    return Frame(cl, -1, base_pointer)

File: cpp/vm/advm/frame.cpp
#include "frame.h"

Frame::Frame(AdClosureObject *c, int i) {
    cl = c;
    ip = i;
    basePointer = 0;
}

Frame::Frame(AdClosureObject *c, int i, int bP) {
    cl = c;
    ip = i;
    basePointer = bP;
}

Frame::~Frame() {
    if (cl != NULL) {
        //free_memory_AdObject(cl); // poate ca da, poate ca nu, vedem daca aici e locul de stergere a closure-ului
        // ceva imi spune ca aici trebuie ceva mai complex
    }
}

Instructions Frame::instructions() {
    return cl->fn->instructions;
}

Frame* newFrame(AdClosureObject *cl, int basePointer) {
    return new Frame(cl, -1, basePointer);
}